From: Daniel Wagner <dwagner@suse.de>
Date: Mon, 12 Oct 2020 10:45:53 +0200
Subject: [PATCH] qla2xxx: Return EBUSY on fcport deletion
Patch-mainline: Submitted - 2020-10-12 - scsi ML
References: bsc#1171688

When the fcport is about to be deleted we should return EBUSY instead
of ENODEV. Only for EBUSY the request will be requeued in a multipath
setup.

Also in case we have a valid qpair but the firmware has not yet
started return EBUSY to avoid dropping the request.

Signed-off-by: Daniel Wagner <dwagner@suse.de>
---

 drivers/scsi/qla2xxx/qla_nvme.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/drivers/scsi/qla2xxx/qla_nvme.c
+++ b/drivers/scsi/qla2xxx/qla_nvme.c
@@ -555,8 +555,11 @@ static int qla_nvme_post_cmd(struct nvme
 
 	fcport = qla_rport->fcport;
 
-	if (!qpair || !fcport || (qpair && !qpair->fw_started) ||
+	if ((qpair && !qpair->fw_started) ||
 	    (fcport && fcport->deleted))
+		return -EBUSY;
+
+	if (!qpair || !fcport)
 		return -ENODEV;
 
 	vha = fcport->vha;
